cmd/utils.go

```diff
@@ -2439,3 +2450,226 @@ func generateCodewindJSON(log *LoggingConfig, indexYaml IndexYaml, indexFilePath
+}
+
+// write to the project.yaml file
+func (p *ProjectFile) writeFile(path string) error {
+	data, err := yaml.Marshal(p)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(path, data, 0644)
+}
+
+// check if project.yaml file had a project with given id
+func (p *ProjectFile) hasID(id string) bool {
+	for _, pf := range p.Projects {
+		if id == pf.ID {
+			return true
+		}
+	}
+	return false
+}
+
+// get project from project.yaml with given id
+func (p *ProjectFile) getProject(id string) *ProjectEntry {
+	for _, pf := range p.Projects {
+		if id == pf.ID {
+			return pf
+		}
+	}
+	return nil
+}
+
+// get all project entries from project.yaml
+func (p *ProjectFile) getProjects(rootConfig *RootCommandConfig) (*ProjectFile, error) {
+	var fileLocation = getProjectYamlFile(rootConfig)
+	projectReader, err := ioutil.ReadFile(fileLocation)
+	if err != nil {
+		return nil, err
+	}
+	err = yaml.Unmarshal(projectReader, p)
+	if err != nil {
+		return nil, errors.Errorf("Failed to parse project file %v", err)
+	}
+	return p, nil
+}
+
+// create unique project id for .appsody-config.yaml
+func generateIDHash(config *RootCommandConfig) string {
+	var password = time.Now().Format("2006-01-02 15:04:05 -0700 MST")
+	hash, _ := bcrypt.GenerateFromPassword([]byte(password), bcrypt.MinCost)
+
+	id := hex.EncodeToString(hash)
+	if len(id) > 100 {
+		config.Debug.Logf("Successfully generated ID: %s", id[0:100])
+		return id[0:100]
+	}
+	config.Debug.Logf("Successfully generated ID: %s", id)
+	return id
+}
+
+// add new project entry to ~/.appsody/project.yaml
+func addNewProject(ID string, config *RootCommandConfig) error {
+	projectDir, err := getProjectDir(config)
+	if err != nil {
+		return err
+	}
+	var projectFile ProjectFile
+	fileLocation := getProjectYamlFile(config)
+
+	_, err = projectFile.getProjects(config)
+	if err != nil {
+		return err
+	}
+
+	var newEntry = ProjectEntry{
+		ID:   ID,
+		Path: projectDir,
+	}
+	projectFile.add(&newEntry)
+	err = projectFile.writeFile(fileLocation)
+	if err != nil {
+		return errors.Errorf("Failed to write file to repository location: %v", err)
+	}
+	config.Info.Logf("Successfully added your project to %s", getProjectYamlFile(config))
+	return nil
+}
+
+// get APPSODY_DEPS environment variable and split it into and array
+func getDepVolumeArgs(config *RootCommandConfig) ([]string, error) {
+	stackDeps, envErr := GetEnvVar("APPSODY_DEPS", config)
+	if envErr != nil {
+		return nil, envErr
+	}
+	if stackDeps == "" {
+		config.Warning.log("The stack image does not contain APPSODY_DEPS")
+		return nil, nil
+	}
+	volumeArgs := strings.Split(stackDeps, ";")
+	return volumeArgs, nil
+}
+
+// create unique name for APPSODY_DEPS volumes
+func generateVolumeName(config *RootCommandConfig) string {
+	projectName, perr := getProjectName(config)
+	if perr != nil {
+		if _, ok := perr.(*NotAnAppsodyProject); !ok {
+			config.Error.logf("Error occurred retrieving project name... exiting: %s", perr)
+			os.Exit(1)
+		}
```

Comment:
Unsure how else to do this as we want to distinguish that the error from `getProjectDir()` is not happening because it is not an Appsody project. I've cleaned up some other cases with empty if statements, other than abstracting the code block to another function, not sure what else we could do..

Reply:
Thanks!